tests/scripts/depends.py

```diff
@@ -0,0 +1,557 @@
+    for key, value in sorted(config_settings.items()):
+        if value is not False:
+            continue
+        for dep in REVERSE_DEPENDENCIES.get(key, []):
+            config_settings[dep] = False
+
+class BaseDomain: # pylint: disable=too-few-public-methods, unused-argument
+    """A base class for all domains."""
+    def __init__(self, symbols, commands, exclude):
+        """Initialize the jobs container"""
+        self.jobs = []
+
+class ExclusiveDomain(BaseDomain): # pylint: disable=too-few-public-methods
+    """A domain consisting of a set of conceptually-equivalent settings.
+Establish a list of configuration symbols. For each symbol, run a test job
+with this symbol set and the others unset."""
+    def __init__(self, symbols, commands, exclude=None):
+        """Build a domain for the specified list of configuration symbols.
+The domain contains a set of jobs that enable one of the elements
+of symbols and disable the others.
+Each job runs the specified commands.
+If exclude is a regular expression, skip generated jobs whose description
+would match this regular expression."""
+        super().__init__(symbols, commands, exclude)
+        base_config_settings = {}
+        for symbol in symbols:
+            base_config_settings[symbol] = False
+        for symbol in symbols:
+            description = symbol
+            if exclude and re.match(exclude, description):
+                continue
+            config_settings = base_config_settings.copy()
+            config_settings[symbol] = True
+            handle_exclusive_groups(config_settings, symbol)
+            turn_off_dependencies(config_settings)
+            job = Job(description, config_settings, commands)
+            self.jobs.append(job)
+
+class ComplementaryDomain(BaseDomain): # pylint: disable=too-few-public-methods
+    """A domain consisting of a set of loosely-related settings.
+Establish a list of configuration symbols. For each symbol, run a test job
+with this symbol unset."""
+    def __init__(self, symbols, commands, exclude=None):
+        """Build a domain for the specified list of configuration symbols.
+Each job in the domain disables one of the specified symbols.
+Each job runs the specified commands."""
+        super().__init__(symbols, commands, exclude)
+        for symbol in symbols:
+            description = '!' + symbol
+            if exclude and re.match(exclude, description):
+                continue
+            config_settings = {symbol: False}
+            turn_off_dependencies(config_settings)
+            job = Job(description, config_settings, commands)
+            self.jobs.append(job)
+
+class DualDomain(ExclusiveDomain, ComplementaryDomain): # pylint: disable=too-few-public-methods
+    """A domain that contains both the ExclusiveDomain and BaseDomain tests"""
+    def __init__(self, symbols, commands, exclude=None):
+        super().__init__(symbols=symbols, commands=commands, exclude=exclude)
```

Comment:
I must admit I don't know off-hand how `super()` behaves with multiple inheritance. I mean, this is probably the most elegant way of expressing this, but I would appreciate a comment about it (perhaps with a pointer to the relevant documentation). See our [pending scripts coding standards](https://github.com/Mbed-TLS/mbedtls-docs/pull/51/files#diff-bdc93beaa341697a5a4b7625d59c0ce3855453745086e9eb387846b01521b99eR92)

Reply:
The assignments themselves weren't required. What exactly would you like to clarify? The reason of each class having the same list of constructor arguments?

Reply:
> Edit: right, I can add some information that both of the parent classes will build their jobs respectively when __init__ is called :)

Yes, that's the bit I wanted. Plus the fact that `jobs` won't be reset between the two (I assume that's the reason why `BaseDomain` was created, so that `self.jobs = []` is only done once).